### ISSUE TITLE
#[5470]: 强化学习资产配置 | RL Asset Allocation

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4354]
**标签**: quantitative-strategy, reinforcement-learning, asset-allocation, portfolio-optimization

### 摘要
使用强化学习（PPO、A2C）进行动态资产配置，在多资产组合中学习最优交易决策。

### 算法描述

**核心原理**

强化学习（Reinforcement Learning）通过与环境交互学习最优策略，无需手工设计特征。在资产配置中，智能体（Agent）观察市场状态（资产价格、收益率、技术指标），输出持仓权重action，接受回报（回报率）并更新策略网络。

**数学基础**

1. **MDP建模**：
   - 状态空间：$S_t = (r_{t-1}, p_{t-1}, volume_{t-1}, \ldots)$
   - 动作空间：$A_t = (w_1, w_2, \ldots, w_n)$ 资产权重，满足 $\sum w_i = 1$
   - 回报函数：$R_t = \sum_{i=1}^{n} w_i \cdot r_{i,t}$

2. **PPO（Proximal Policy Optimization）**：
   $$L^{CLIP}(\theta) = \mathbb{E}_t \left[ \min\left(r_t(\theta) A_t, \text{clip}(r_t(\theta), 1-\epsilon, 1+\epsilon) A_t\right) \right]$$
   其中 $r_t(\theta) = \frac{\pi_\theta(a_t|s_t)}{\pi_{\theta_{old}}(a_t|s_t)}$ 为概率比。

3. **A2C（Advantage Actor-Critic）**：
   - Actor：$\nabla_{\theta} \log \pi_\theta(a|s) \cdot A(s,a)$
   - Critic：$L(\phi) = \mathbb{E}[(V_\phi(s) - G_t)^2]$
   其中 $A(s,a) = Q(s,a) - V(s)$ 为优势函数。

4. **投资组合回报**：
   $$R_{portfolio} = \sum_{i=1}^{n} w_i \cdot \frac{P_{i,t} - P_{i,t-1}}{P_{i,t-1}}$$

**计算流程**

1. **环境构建**：创建gym风格环境，模拟资产价格序列
2. **状态提取**：计算收益率、技术指标作为状态特征
3. **策略网络**：输出资产权重（softmax确保和为1）
4. **价值网络**：估计状态价值函数
5. **策略更新**：PPO/A2C算法更新网络参数
6. **回测评估**：在样本外数据验证策略表现

### Python代码示例

```python
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
import torch.optim as optim
from collections import deque
import random


class PortfolioEnv:
    """投资组合强化学习环境"""
    
    def __init__(self, prices_df, initial_balance=100000, transaction_cost=0.001):
        self.prices = prices_df
        self.n_assets = len(prices_df.columns)
        self.initial_balance = initial_balance
        self.transaction_cost = transaction_cost
        self.reset()
    
    def reset(self):
        self.current_step = 0
        self.balance = self.initial_balance
        self.positions = np.zeros(self.n_assets)
        self.portfolio_value = [self.initial_balance]
        return self._get_state()
    
    def _get_state(self):
        """获取当前状态"""
        if self.current_step < 1:
            returns = np.zeros(self.n_assets)
        else:
            returns = (self.prices.iloc[self.current_step] / self.prices.iloc[self.current_step - 1] - 1).values
        
        state = np.concatenate([
            returns,
            self.positions / (self.portfolio_value[-1] + 1e-8) * 100,
            [self.balance / (self.portfolio_value[-1] + 1e-8)]
        ])
        return state.astype(np.float32)
    
    def step(self, action):
        """执行动作，返回下一状态、奖励、是否结束"""
        action = np.clip(action, 0, 1)
        action = action / (action.sum() + 1e-8)
        
        target_value = self.portfolio_value[-1] * action
        current_value = self.positions * self.prices.iloc[self.current_step].values
        
        trades = target_value - current_value
        transaction_cost = np.abs(trades).sum() * self.transaction_cost
        
        self.balance -= transaction_cost
        
        self.positions = target_value / (self.prices.iloc[self.current_step].values + 1e-8)
        
        self.current_step += 1
        
        if self.current_step >= len(self.prices) - 1:
            done = True
        else:
            done = False
        
        new_portfolio_value = (self.positions * self.prices.iloc[min(self.current_step, len(self.prices)-1)].values).sum() + self.balance
        self.portfolio_value.append(new_portfolio_value)
        
        reward = (new_portfolio_value - self.portfolio_value[-2]) / self.portfolio_value[-2]
        state = self._get_state()
        
        return state, reward, done, {}
    
    def render(self):
        print(f"Step: {self.current_step}, Portfolio Value: {self.portfolio_value[-1]:.2f}")


class ActorNetwork(nn.Module):
    """Actor网络：输出动作（资产权重）"""
    
    def __init__(self, state_dim, action_dim, hidden_dim=256):
        super().__init__()
        self.network = nn.Sequential(
            nn.Linear(state_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, action_dim),
            nn.Softmax(dim=-1)
        )
    
    def forward(self, state):
        return self.network(state)


class CriticNetwork(nn.Module):
    """Critic网络：输出状态价值"""
    
    def __init__(self, state_dim, hidden_dim=256):
        super().__init__()
        self.network = nn.Sequential(
            nn.Linear(state_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, 1)
        )
    
    def forward(self, state):
        return self.network(state)


class A2CAgent:
    """A2C强化学习智能体"""
    
    def __init__(self, state_dim, action_dim, lr_actor=0.001, lr_critic=0.002, gamma=0.99):
        self.gamma = gamma
        self.actor = ActorNetwork(state_dim, action_dim)
        self.critic = CriticNetwork(state_dim)
        self.actor_optimizer = optim.Adam(self.actor.parameters(), lr=lr_actor)
        self.critic_optimizer = optim.Adam(self.critic.parameters(), lr=lr_critic)
        self.memory = deque(maxlen=10000)
    
    def select_action(self, state):
        state_tensor = torch.FloatTensor(state).unsqueeze(0)
        probs = self.actor(state_tensor)
        action_dist = torch.distributions.Categorical(probs)
        action = action_dist.sample()
        return action.item(), probs[0, action.item()]
    
    def select_action_continuous(self, state):
        state_tensor = torch.FloatTensor(state).unsqueeze(0)
        probs = self.actor(state_tensor).squeeze()
        return probs.detach().numpy()
    
    def update(self, states, actions, rewards, dones, next_states):
        states_tensor = torch.FloatTensor(states)
        next_states_tensor = torch.FloatTensor(next_states)
        actions_tensor = torch.LongTensor(actions)
        rewards_tensor = torch.FloatTensor(rewards)
        dones_tensor = torch.FloatTensor(dones)
        
        values = self.critic(states_tensor).squeeze()
        next_values = self.critic(next_states_tensor).squeeze()
        
        targets = rewards_tensor + self.gamma * next_values * (1 - dones_tensor)
        advantages = targets - values
        
        critic_loss = advantages.pow(2).mean()
        self.critic_optimizer.zero_grad()
        critic_loss.backward()
        self.critic_optimizer.step()
        
        actor_loss = -advantages.mean()
        self.actor_optimizer.zero_grad()
        actor_loss.backward()
        self.actor_optimizer.step()
        
        return advantages.mean().item()


class ReplayBuffer:
    """经验回放缓冲区"""
    
    def __init__(self, capacity=10000):
        self.buffer = deque(maxlen=capacity)
    
    def push(self, state, action, reward, next_state, done):
        self.buffer.append((state, action, reward, next_state, done))
    
    def sample(self, batch_size):
        batch = random.sample(self.buffer, batch_size)
        states, actions, rewards, next_states, dones = zip(*batch)
        return (
            np.array(states),
            np.array(actions),
            np.array(rewards),
            np.array(next_states),
            np.array(dones)
        )
    
    def __len__(self):
        return len(self.buffer)


def generate_synthetic_prices():
    """生成模拟价格数据"""
    np.random.seed(42)
    n_days = 500
    n_assets = 5
    
    initial_prices = np.array([100, 150, 200, 80, 120])
    prices = pd.DataFrame()
    
    for i in range(n_assets):
        drift = np.random.uniform(0.0001, 0.0003)
        volatility = np.random.uniform(0.01, 0.025)
        returns = np.random.randn(n_days) * volatility + drift
        price_series = initial_prices[i] * np.exp(np.cumsum(returns))
        prices[f'asset_{i}'] = price_series
    
    return prices


def train_agent(env, agent, n_episodes=100, batch_size=32):
    """训练智能体"""
    results = []
    
    for episode in range(n_episodes):
        state = env.reset()
        total_reward = 0
        states, actions, rewards, next_states, dones = [], [], [], [], []
        
        while True:
            action = agent.select_action_continuous(state)
            
            next_state, reward, done, _ = env.step(action)
            
            states.append(state)
            actions.append(np.argmax(action))
            rewards.append(reward)
            next_states.append(next_state)
            dones.append(done)
            
            state = next_state
            total_reward += reward
            
            if done:
                if len(states) >= batch_size:
                    agent.update(states, actions, rewards, next_states, dones)
                
                results.append({
                    'episode': episode,
                    'total_reward': total_reward,
                    'portfolio_value': env.portfolio_value[-1]
                })
                break
        
        if (episode + 1) % 10 == 0:
            avg_reward = np.mean([r['total_reward'] for r in results[-10:]])
            print(f"Episode {episode+1}, Avg Reward: {avg_reward:.4f}, Portfolio: {env.portfolio_value[-1]:.2f}")
    
    return results


def main():
    print("=" * 60)
    print("强化学习资产配置策略演示")
    print("=" * 60)
    
    prices = generate_synthetic_prices()
    print(f"\n资产数量: {len(prices.columns)}")
    print(f"数据天数: {len(prices)}")
    
    env = PortfolioEnv(prices, initial_balance=100000)
    state_dim = env.n_assets * 2 + 1
    action_dim = env.n_assets
    
    agent = A2CAgent(state_dim, action_dim, lr_actor=0.001, lr_critic=0.002)
    
    print("\n训练智能体...")
    results = train_agent(env, agent, n_episodes=50)
    
    print("\n训练后测试:")
    state = env.reset()
    while True:
        action = agent.select_action_continuous(state)
        next_state, reward, done, _ = env.step(action)
        state = next_state
        if done:
            break
    
    print(f"最终组合价值: {env.portfolio_value[-1]:.2f}")
    print(f"总回报率: {(env.portfolio_value[-1] / 100000 - 1) * 100:.2f}%")


if __name__ == "__main__":
    main()
```

### 数据来源
- Reddit: r/quant, r/algorithmictrading, r/Finance
- 小红书: 量化投资、理财技巧
- 知乎: 量化交易、机器学习交易
- 其他: Investopedia, Quantopian, arXiv
